checkmate/models/db/mixins.py

```diff
@@ -0,0 +1,117 @@
+"""Mixins to enhance model objects."""
+
+import sqlalchemy as sa
+from sqlalchemy.dialects.postgresql import insert
+from zope.sqlalchemy import mark_changed
+
+
+class HashMatchMixin:
+    """A mixin for models which want to support hash based URL comparisons.
+
+    Your subclass must have a column like this:
+        hash = HashMatchMixin.hash_column()
+    """
+
+    @staticmethod
+    def hash_column(unique=False):
+        """Return a column suitable for hashing which you must call "hash"."""
+
+        # The C collation here allows the B-Tree default indexing type to be
+        # used for prefix matching. This makes searching for 'ab2d23d45a%'
+        # very quick: https://www.postgresql.org/docs/11/indexes-types.html
+        return sa.Column(
+            sa.String(collation="C"), nullable=False, index=True, unique=unique
+        )
```

Comment:
I think this may be able to be just an attribute:

```python
hash = sa.Column(sa.String(collation="C"), nullable=False, index=True)
```

So that subclasses would inherit a non-unique hash column by default. And then if a subclass wanted to make the column unique, it could add its own unique constraint like this:

```python
class SubClass(Base, HashMatchMixin):
    __table_args__ = (UniqueConstraint("hash"))
```

The `@staticmethod` way should also work fine as far as I know, though? I don't see why not

Reply:
Mixins in SQLAlchemy look like they can get pretty thorny, FYI: https://docs.sqlalchemy.org/en/14/orm/declarative_mixins.html

Reply:
Yeah I kind of like it, but I also kind of hate it because you can't look at the original file and know how many columns it has etc.